### PR TITLE
Add support for XDG_CONFIG_HOME

### PR DESCRIPTION
**Pull request by 0az**
Implement support for `XDG_CONFIG_HOME`.

To avoid breaking changes, we use locations in this order:
- `$B2_ACCOUNT_INFO_ENV_VAR`: If set
- `~/.b2_account_info`: If exists
- `$XDG_CONFIG_HOME/b2/account_info`: If set
- `~/.b2_account_info`: If no path takes priority

### CODE REVIEW

**Comment by ppolewicz** [comment]
We used to be able to pass a file path to the constructor of SqliteAccountInfo, but you removed it. Please restore it.

Also, could you briefly explain why are you proposing this change?

**Comment by 0az** [comment]
> We used to be able to pass a file path to the constructor of SqliteAccountInfo, but you removed it. Please restore it.

Passing in the file path still works: I just forgot to add it to the explanatory text and the docstring. I have since done the latter.

> Also, could you briefly explain why are you proposing this change

I'm proposing the change so I can remove another dotfile from `$HOME` and add b2sdk/b2 cli to the list of applications that support the XDG spec. If I'm not mistaken, version 1.1.0 should go out soonish, and will be used in the next version of the b2 cli.

**Comment by ppolewicz** [comment]
I think it is a good change. Could you please add a unit test to make sure that we don't hit a regression on this new functionality in the future?

**Comment by 0az** [comment]
I added unit tests for two cases:
- `XDG_CONFIG_HOME`
- `B2_ACCOUNT_INFO_ENV_VAR` and `XDG_CONFIG_HOME` are set.

I'm not sure how to test the third interaction, where `XDG_CONFIG_HOME` is set, but a database already exists. This may require changing the `file_name` kwarg to override `B2_ACCOUNT_INFO_DEFAULT_FILE` instead of overriding everything.

A hermetic testing environment and compatibility with existing code is ensured by modifying `_make_sqlite_account_info` to take an `env` kwarg. If it is unset or falsy, it will instantiate SqliteAccountInfo with an empty env. If it is truthy, it will override the usual `file_name` behavior.

File: b2sdk/account_info/sqlite_account_info.py
Comment (bwbeach) [inline b2sdk/account_info/sqlite_account_info.py:50]: I think this code works, but the logic would be clearer if it mirrored the structure of the comment, above:

```
if file_name:
    use it
elif B2_ACCOUNT_INFO_ENV_VAR in env:
    use it
elif default file exists:
    use it
elif XDG_CONFIG_HOME in env:
    use it
else:
    use default path
```

File: test/v1/test_account_info.py
Comment (ppolewicz) [inline test/v1/test_account_info.py:302]: please add a trailing `,`  in this line so that `yapf` reliably guards the format of the multi-line statement

File: test/v1/test_account_info.py
Comment (ppolewicz) [inline test/v1/test_account_info.py:309]: `tempfile.TemporaryDirectory` was added in 3.2, but we still support 2.7, but we do have [b2sdk.utils.TempDir](https://github.com/Backblaze/b2-sdk-python/blob/43b42e7562ec82c8baefd2614b1be6aefbfb0e0e/b2sdk/utils.py#L239) context manager. Please switch to it.

File: test/v1/test_account_info.py
Comment (ppolewicz) [inline test/v1/test_account_info.py:313]:
```suggestion
            assert expected_path == actual_path, 'Actual path %s is not equal to $XDG_CONFIG_HOME/b2/account_info' % (actual_path,)
```

File: test/v1/test_account_info.py
Comment (ppolewicz) [inline test/v1/test_account_info.py:330]:
```suggestion
            assert expected_path == actual_path, 'Actual path %s is not equal to %s' % (
                actual_path,
                B2_ACCOUNT_INFO_ENV_VAR,
            )
```
Most of the code uses % for string formatting, I've only now discovered that one new module uses `.format()` and I have filed an issue to fix it. We should keep to the `%`, at least until we drop 3.5, so that the code is consistent across the project.

File: test/v1/test_account_info.py
Comment (ppolewicz) [inline test/v1/test_account_info.py:317]: I think you can use `self.assertEquals` here, it's a bit cleaner

Comment (ppolewicz) [reply test/v1/test_account_info.py:317]: in the next test too

File: b2sdk/account_info/sqlite_account_info.py
Comment (ppolewicz) [inline b2sdk/account_info/sqlite_account_info.py:65]: this was actually not resolved, will fail on 2.7

File: test/v1/test_account_info.py
Comment (ppolewicz) [inline test/v1/test_account_info.py:305]: you have removed the `env` argument from `SqliteAccountInfo` constructor, but it remained here in tests, causing all tests of `SqliteAccountInfo` to fail. Could you please fix it and run `./pre-commit.sh` to make sure all tests pass before we merge it in?

Also, could you please add `Add support for XDG_CONFIG_HOME (0az)` to changelog section in readme?

**Comment by 0az** [comment]
I'm really sorry about that.

I probably should have mentioned this earlier, but for some unknown reason that I don't have time to debug after midnight, the pre-commit shell script fails with a RuntimeError: "An attempt has been made to start a new process before the current process has finished its bootstrapping phase." Logs are in this gist: https://gist.github.com/0az/234a94a123e15259f3779661feed69e6. Manually running nosetests still works as a workaround, though.

I also modified the test class to include a temporary directory acting as HOME.

**Comment by ppolewicz** [comment]
@mlech-reef did we accidentally enable parallel testing in nosetests on Windows? I don't think that ever worked, now appveyor can't run the tests and our contributor here also has issues. Travis is absent from this PR for some reason... Do you have any clue?

**Comment by mlech-reef** [comment]
Travis is triggered properly, but for some reason, it's not showing here. I've seen this issue in other PRs. We should investigate it.

**Comment by mlech-reef** [comment]
> @mlech-reef did we accidentally enable parallel testing in nosetests on Windows? I don't think that ever worked, now appveyor can't run the tests and our contributor here also has issues.

I don't recall anything like that. AppVeyor fails not because of the parallel runs but because 2 new tests are failing with:
`PermissionError: [WinError 32] The process cannot access the file because it is being used by another process`
It's probably because `TempDir` uses `shutils.rmtree` on a directory with sqlite db file created inside.
It's similar to the problem described here: https://stackoverflow.com/questions/1213706/what-user-do-python-scripts-run-as-in-windows
Windows and python sometimes don't get along :)

File: b2sdk/account_info/sqlite_account_info.py
Comment (ppolewicz) [inline b2sdk/account_info/sqlite_account_info.py:39]: I am reading this docstring over and over and I think apiver doesn't let us change the default location to use `XDG_CONFIG_HOME` in v1. Someone may have written a program that relies on the promise that when the filename was not passed and `B2_ACCOUNT_INFO` was not set, then the db file will land in `~/.b2_account_info`. After running something, the program might clean the environment by removing that file.

If we make this change and `XDG_CONFIG_HOME` is defined, sdk user would not expect the configuration of the end user to be suddenly leaked (as it is not cleaned up anymore), possibly allowing access to the following user.

I think this is a good change and we should introduce it, but that should happen with the release of `from b2sdk.v2 import *`, otherwise we risk introducing a breaking change, while the purpose of apiver is to avoid it unless fixing a bug (which we are not doing here) which would change behavior because the old behavior was just wrong.

Furthermore, we should separate the documentation of the interface from the documentation of the current behavior, so that this can be altered in the future, by mentioning that the user should not define of this algorithm staying exactly the same, but should use `self.filename` to determine the actually determined location of the file in case it needs to be accessed later.

We did not have a compelling reason to release v2, but this is a good one.

@bwbeach @0az do you agree with my viewpoint? Shall we move it to v2?

Comment (0az) [reply b2sdk/account_info/sqlite_account_info.py:39]: Good catch: I didn't actually think of that case.

Sounds good to me.

Comment (ppolewicz) [reply b2sdk/account_info/sqlite_account_info.py:39]: @0az then would you please adjust the docstring here? We'll hold off merging this until b2sdk 1.1.0 and b2cli 2.0.0 are shipped, then we'll increment apiver to v2 and finally we'll merge your thing (though a final adjustment will be necessary to provide backwards compatibility)

Comment (0az) [reply b2sdk/account_info/sqlite_account_info.py:39]: Another question for v2: should SqliteAccountInfo support `:memory:` or anonymous (empty string) databases?

https://www.sqlite.org/c3ref/open.html
> If the filename is ":memory:", then a private, temporary in-memory database is created for the connection. This in-memory database will vanish when the database connection is closed. Future versions of SQLite might make use of additional special filenames that begin with the ":" character. It is recommended that when a database filename actually does begin with a ":" character you should prefix the filename with a pathname such as "./" to avoid ambiguity.
> 
> If the filename is an empty string, then a private, temporary on-disk database will be created. This private database will be automatically deleted as soon as the database connection is closed.

File: b2sdk/account_info/sqlite_account_info.py
Comment (0az) [inline b2sdk/account_info/sqlite_account_info.py:41]:
```suggestion
        The exact algorithm used to determine the location of the database file is not API in any sense.
        If the location of the database file is required (for cleanup, etc), do not assume a specific resolution:
        instead, use ``self.filename`` to get the actual resolved location.

        SqliteAccountInfo currently checks locations in the following order:
```
@ppolewicz How does this look?

File: b2sdk/account_info/sqlite_account_info.py
Comment (ppolewicz) [reply b2sdk/account_info/sqlite_account_info.py:39]: It looks like that could replace `v` with one line of code. I like less code.

File: b2sdk/account_info/sqlite_account_info.py
Comment (ppolewicz) [reply b2sdk/account_info/sqlite_account_info.py:41]: that's way better, thanks

**Comment by mpnowacki-reef** [comment]
@0az I have submitted a PR that merges master into your `xdg` branch and moves your changes along with tests to v2: https://github.com/0az/b2-sdk-python/pull/1/
We're planning to wrap up v2 by the end of this week, would you be willing to finish this PR?

**Comment by 0az** [comment]
Thanks! Was there anything else that I needed to do? It's been a while. 🙃

File: b2sdk/v1/account_info.py
Comment (ppolewicz) [inline b2sdk/v1/account_info.py:95]: where is `logger` initialized?

**Review by ppolewicz — CHANGES_REQUESTED** [review]

**Comment by ppolewicz** [comment]
This PR seems to contain v2 changes. Can we somehow make it clean? @0az would you mind if we took your changes and rebased them to slide it into v2 at the appropriate moment? The v2 change is a little bit chaotic and I don't want to keep pulling you into this.

**Comment by 0az** [comment]
> This PR seems to contain v2 changes. Can we somehow make it clean?

Michał sent in a PR with the v2 changes, so I merged that in. I can force push that off of the branch, if you want.

> @0az would you mind if we took your changes and rebased them to slide it into v2 at the appropriate moment? The v2 change is a little bit chaotic and I don't want to keep pulling you into this.

This is also fine.

File: b2sdk/v1/account_info.py
Comment (0az) [reply b2sdk/v1/account_info.py:95]: It's imported from here:

https://github.com/Backblaze/b2-sdk-python/blob/632b13b74aa584a9c962a9b85f4630b8373cba1d/b2sdk/account_info/sqlite_account_info.py#L23

Comment (ppolewicz) [reply b2sdk/v1/account_info.py:95]: ok, but here in `b2sdk/v1/account_info.py` there is no `logger`, so it will crash. Logger initialization should be added.

Comment (0az) [reply b2sdk/v1/account_info.py:95]: Sorry, bad wording. The initialization happens there, and there's an import in this file *for* that logger.

It probably should get initialized here anyway, though.

Comment (ppolewicz) [reply b2sdk/v1/account_info.py:95]: `from b2sdk.account_info.sqlite_account_info import logger` ohhh. I don't think we should do that, that is super confusing. We should initialize it locally, right?

**Comment by mpnowacki-reef** [comment]
@0az I've come to clean up the mess I made. I squashed all your commits on top of the current master and created a new PR (https://github.com/Backblaze/b2-sdk-python/pull/235/). I have made you the author of the first commit, in order not to steal your thunder. Is that okay with you?

**Comment by 0az** [comment]
Yeah, 235 looks good. I'll close this PR, then.

**Pull request closed without merging**
